src/main/java/bio/terra/common/retry/SamRetry.java

```diff
@@ -0,0 +1,136 @@
+  private final Instant operationTimeout;
+
+  // How long to wait between retries.
+  private Duration retryDuration;
+
+  private static final ObjectMapper objectMapper = new ObjectMapper();
+
+  SamRetry() {
+    this.operationTimeout = now().plus(OPERATION_TIMEOUT);
+    this.retryDuration = INITIAL_WAIT;
+  }
+
+  protected SamRetry(Duration timeout) {
+    this.operationTimeout = now().plus(timeout);
+    this.retryDuration = INITIAL_WAIT;
+  }
+
+  @FunctionalInterface
+  public interface SamVoidFunction {
+    void apply() throws ApiException, InterruptedException;
+  }
+
+  @FunctionalInterface
+  public interface SamFunction<R> {
+    R apply() throws ApiException, InterruptedException;
+  }
+
+  public static <T> T retry(SamFunction<T> function) throws SamApiException {
+    SamRetry samRetry = new SamRetry();
+    return samRetry.perform(function);
+  }
+
+  // Version of the retry function for testing so that we can override timeout
+  protected static <T> T retry(SamFunction<T> function, Duration timeout) throws SamApiException {
+    SamRetry samRetry = new SamRetry(timeout);
+    return samRetry.perform(function);
+  }
+
+  public static void retry(SamVoidFunction function) throws SamApiException {
+    SamRetry samRetry = new SamRetry();
+    samRetry.performVoid(function);
+  }
+
+  // Version of the retry function for testing so that we can override timeout
+  protected static void retry(SamVoidFunction function, Duration timeout) throws SamApiException {
+    SamRetry samRetry = new SamRetry(timeout);
+    samRetry.performVoid(function);
+  }
+
+  private <T> T perform(SamFunction<T> function) throws SamApiException {
+    while (true) {
+      try {
+        return function.apply();
+      } catch (ApiException ex) {
+        SamApiException samApiException = SamApiException.createSamApiException(ex);
+        if (isRetryable(samApiException)) {
+          logger.info("SamRetry: caught retry-able exception: ", ex);
+          sleepOrTimeoutBeforeRetrying(samApiException);
+        } else {
+          throw samApiException;
+        }
+      } catch (Exception ex) {
+        throw new InternalServerErrorException("Unexpected exception type: " + ex.toString(), ex);
+      }
+    }
+  }
+
+  private boolean isRetryable(SamApiException samApiException) {
+    return (samApiException.getStatusCode() == HttpStatus.INTERNAL_SERVER_ERROR);
+  }
+
+  private void performVoid(SamVoidFunction function) throws SamApiException {
+    perform(
+        () -> {
+          function.apply();
+          return null;
+        });
+  }
+
+  /**
+   * Given an exception from Sam, either timeout and rethrow the error from Sam
+   * or sleep for retryDuration. If the thread times out while sleeping, throw the initial
+   * exception.
+   * @param samApiException The error Sam threw
+   * @throws SamApiException
+   */
+  private void sleepOrTimeoutBeforeRetrying(SamApiException samApiException)
```

Comment:
nit: maybe rename the argument to `previousException` or `initialException`?

Reply:
Done.